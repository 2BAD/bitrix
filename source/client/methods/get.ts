import { GotInstance, GotJSONFn } from 'got'
import { stringify as toQuery } from 'qs'
import { GetParams, GetPayload, GettableMethod } from '../types'

export type Get = <P>(method: GettableMethod, params?: GetParams) => Promise<GetPayload<P>>

/**
 * Dispatches a request with specified method and params.
 * Should be used for getting single items only, otherwise payload will be wrongly typed.
 */
export default (got: GotInstance<GotJSONFn>): Get => {
  const get: Get = <P>(method: GettableMethod, params?: GetParams): Promise<GetPayload<P>> =>
    got.get(method, { query: toQuery(params) })
        .then(({ body }) => body as GetPayload<P>)

  return get
}
