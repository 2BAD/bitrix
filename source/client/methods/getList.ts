import { GotInstance, GotJSONFn } from 'got'
import { stringify as toQuery } from 'qs'
import { ListableMethod, ListParams, ListPayload } from '../../types'

export const handleGetListPayload = <P>(payload: ListPayload<P>) => {
  // tslint:disable-next-line no-if-statement
  if (payload.error) {
  // tslint:disable-next-line no-throw
    throw new Error(
      `[get] failed to get the resource: ${payload.error}.`
    )
  }

  return payload
}

// @todo `getList` is temporary workaround. Bitrix will return different signature
//       depending on `get` and `list` methods. Until we can automatically
//       map those types based on methods and thus infer output types, we need this helper
export default ({ get }: GotInstance<GotJSONFn>) =>
  <P>(method: ListableMethod, params?: ListParams): Promise<ListPayload<P>> =>
    get(method, { query: toQuery(params) })
      .then(({ body }) => handleGetListPayload(body as ListPayload<P>))
