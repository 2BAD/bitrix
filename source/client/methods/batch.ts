import { GotJSONFn } from 'got'
import chunk from 'lodash.chunk'
import fromPairs from 'lodash.frompairs'
import { stringify as toQuery } from 'qs'
import { Commands } from '../../commands'
import { Method, MethodPayloadType } from '../../methods'
import { BatchPayload } from '../../payloads'
import isArray from '../../utils/isArray'
import { handlePayload } from './call'

export const MAX_COMMANDS_PER_BATCH = 50

/**
 * Split any number of commands into the chunks with max commands per batch in each chunk
 */
export const chunkCommands = <C extends Commands>(
  commands: C,
  size: number = MAX_COMMANDS_PER_BATCH
): ReadonlyArray<C> => {
  const commandsIsArray = isArray(commands)
  const chunks = chunk(Object.entries(commands), size)

  // @todo Generated return type isn't very accurate since it says that it will make array of <C>, while
  //       in fact <C> will be distributed among chunks. That gives wrong hints with named commands and arrays as const.
  return chunks.map(
    (c) => commandsIsArray
      // @todo Figure out how to avoid unsafe casting to <C> here
      //       Without `<C>` generic it won't work with complex functions like `batch`
      ? c.map(([_key, command]) => command) as unknown as C
      : fromPairs(c) as C
  )
}

/**
 * Prepares list of commands to be used in requests to the Bitrix
 * @note We could avoid that function... if only Bitrix API would support posting of the batch commands as
 *       plain objects, like it does with other methods. Instead, it should be a dict or array of string queries
 */
export const prepareCommandsQueries = <C extends Commands, R = { [K in keyof C]: string }>(commands: C): R =>
  Object.entries(commands).reduce((calls, [name, { method, params }]) => {
    const stringifiedParams = params ? `?${toQuery(params)}` : ''

    return {
      ...calls,
      [`cmd[${name}]`]: `${method}${stringifiedParams}`
    }
  // tslint:disable-next-line: no-object-literal-type-assertion
  }, {} as R)

/**
 * Merges list of batch responses into a single batch
 * @todo Generics inference is complicated here and might be not super accurate
 */
export const mergeBatchPayloads = <
  B extends BatchPayload<any>,
  P = B extends BatchPayload<infer U> ? U : never
>(payloads: ReadonlyArray<B>): BatchPayload<P> => {
  const merge = <T extends S, S extends readonly any[] | Record<string, any>>(a: T, b: S): T =>
    isArray(a) && isArray(b)
      // @todo Gotta find way to avoid unsafe casting here
      ? [...a, ...b] as unknown as T
      : ({ ...a, ...b })

  return payloads.reduce((merged, payload) => ({
    result: {
      result: merge(merged.result.result || [], payload.result.result),
      result_error: merge(merged.result.result_error || [], payload.result.result_error),
      result_next: merge(merged.result.result_next || [], payload.result.result_next),
      result_time: merge(merged.result.result_time || [], payload.result.result_time),
      result_total: merge(merged.result.result_total || [], payload.result.result_total)
    },
    time: { ...merged.time, ...payload.time }
  // tslint:disable-next-line: no-object-literal-type-assertion
  }), { result: {}, time: {} } as BatchPayload<P>)
}

export type CommandsPayloads<
  C extends Commands,
  CM = {
    [K in keyof C]: C[K] extends { readonly method: infer M } ? M : never
  }
  > = {
  [K in keyof CM]: MethodPayloadType<CM[K] extends Method ? CM[K] : never>
}

// @todo Using `CommandsWithRequiredParamsTypes` and `Batch2` we could ensure that only supported commands
//       are supplied to the `batch`, but it results in unknown error when assigning `Batch2` to `batch` constant.

// export type CommandsWithRequiredParamsTypes<
//   C extends Commands,
//   CC = {
//     [K in keyof C]: C[K] extends Command ? {
//       method: C[K]['method'], params: MethodParams<C[K]['method']>
//     } : Command
//   },
// > = CC

// export type Batch2 = <
//   C extends Commands & { [K in keyof CC]: CC[K] },
//   CC = CommandsWithRequiredParamsTypes<C>,
//   P = CommandsPayloads<C>
// >(commands: C, commandsPerRequest?: number) => Promise<BatchPayload<P>>

export type Batch = <
  C extends Commands,
  P = CommandsPayloads<C>
>(commands: C, commandsPerRequest?: number) => Promise<BatchPayload<P>>

interface Dependencies {
  readonly get: GotJSONFn
}

/**
 * Dispatches a batch request with specified commands. Will fill figure out payload type based on the Methods.
 * Supports unlimited number of commands. If they do exceed max amount of commands per batch, will dispatch
 * multiple request and merge result into single batch payload
 * @todo For now due to issues with types it won't check `params` of command, so any Method can be used
 *       with any (even wrong) params
 */
export default ({ get }: Dependencies): Batch => {
  const batch: Batch = async <
    C extends Commands,
    P = CommandsPayloads<C>
  >(
    commands: C,
    commandsPerRequest: number = MAX_COMMANDS_PER_BATCH
  ): Promise<BatchPayload<P>> => {
    const call = (c: C) =>
      get(Method.BATCH, { query: prepareCommandsQueries(c) })
        .then(({ body }) => body as BatchPayload<P>)

    const calls = chunkCommands(commands, commandsPerRequest)
      .map(call)

    return Promise.all(calls)
      .then((chunkedPayloads) => {
        const payloads = mergeBatchPayloads(chunkedPayloads)
        return handlePayload(payloads)
      })
  }

  return batch
}
