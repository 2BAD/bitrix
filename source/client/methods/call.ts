import got from 'got'
import { stringify as toQuery } from 'qs'
import { Method, MethodParams, MethodPayload } from '../../methods'
import { BatchPayload, ListPayload, Payload } from '../../payloads'
import isArray from '../../utils/isArray'

/**
 * Checks wether payload have any errors and if it does â€” throws them
 * Bitrix payload types do not provide discriminators, so we're forced to type cast them
 */
export const handlePayload = <P extends Payload<unknown>>(payload: P): P => {
  if ((payload as ListPayload<unknown>).error) {
    throw new Error(
      `[call] failed to get the resource: ${(payload as ListPayload<unknown>).error ?? ''}.`
    )
  }

  if ((payload as BatchPayload<unknown>).result && (payload as BatchPayload<unknown>).result.result_error) {
    const resultErrors = (payload as BatchPayload<unknown>).result.result_error
    const errors = isArray(resultErrors) ? resultErrors : Object.values(resultErrors)

    if (errors.length > 0) {
      // @todo We can give better formatting to display errored commands. But it's not important for now
      throw new Error(`[batch] failed to process. Received errors in ${errors.length} commands:\n${errors.join('\n')}`)
    }
  }

  return payload
}

export type Call = <M extends Method>(method: M, params: MethodParams<M>) => Promise<MethodPayload<M>>

type Dependencies = {
  readonly get: typeof got.get
}

/**
 * Dispatches a request with specified method and params. Will fill figure out payload type based on the Method
 */
export default ({ get }: Dependencies): Call => {
  const call: Call = <M extends Method>(method: M, params: MethodParams<M>): Promise<MethodPayload<M>> =>
    get<MethodPayload<M>>(method, { searchParams: toQuery(params) })
      .then(({ body }) => handlePayload(body))

  return call
}
