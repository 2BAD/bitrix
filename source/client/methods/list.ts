import range from 'lodash.range'
import {
  BatchPayload,
  Command,
  Commands,
  ListableMethod,
  ListParams,
  ListPayload
} from '../../types'
import { Batch } from './batch'
import { GetList } from './getList'

const MAX_ENTRIES_PER_COMMAND = 50

/**
 * Generates required amount of commands to process specified amount of entries
 */
export const fillWithCommands = (
  { method, params }: Command,
  start: number,
  toProcess: number,
  entriesPerCommand: number
): Commands => {
  const requiresCommands = Math.ceil((toProcess - start) / entriesPerCommand)

  return range(0, requiresCommands)
    .map((i) => ({ method, params: { ...params, start: start + (entriesPerCommand * i) } }), {})
}

/**
 * Converts batch payload to a list payload
 */
const batchToListPayload = <P>(payload: BatchPayload<Record<string | number, readonly P[]>>): ListPayload<P> => {
  const { result: { result, result_total, result_error, result_next }, time } = payload

  const flattenResult = Object.entries(result).reduce(
    (flatten, [_key, r]) => !r ? flatten : [...flatten, ...r]
  , [] as readonly P[])

  /**
   * Get the highest value from object or an array if there's any
   */
  const highest = (input: ReadonlyArray<number | undefined> | Record<string, number | undefined>): number | undefined =>
    Object.values(input).reduce(
      (a, b) => a === undefined || b === undefined ? undefined : a > b ? a : b
    , 0)

  return {
    error: Object.values(result_error).join('\n'),
    next: highest(result_next),
    result: flattenResult,
    // @todo Not accurate, we do not care
    time,
    total: highest(result_total) || 0
  }
}

interface Dependencies {
  readonly getList: GetList
  readonly batch: Batch
}

export type List = <P>(method: ListableMethod, params?: ListParams) => Promise<ListPayload<P>>

/**
 * Gets all entries by dispatching required amount of requests
 */
export default ({ getList, batch }: Dependencies): List => {
  const list: List = async <P>(method: ListableMethod, params: ListParams = {}): Promise<ListPayload<P>> => {
    const start = params.start || 0

    const listAll = async () => {
      const batchCommands = fillWithCommands({ method, params }, start, firstCall.total, MAX_ENTRIES_PER_COMMAND)
      const payload = await batch<Record<string | number, readonly P[]>>(batchCommands)

      return batchToListPayload(payload)
    }

    const firstCall = await getList<P>(method, { ...params, start })

    return !firstCall.next ? firstCall : listAll()
  }

  return list
}
