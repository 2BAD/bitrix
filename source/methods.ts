import { Commands } from './commands.js'
import { BatchPayload } from './payloads.js'
import { CompaniesMethods } from './services/companies/methods.js'
import { ContactsMethods } from './services/contacts/methods.js'
import { DealsMethods } from './services/deals/methods.js'
import { LeadsMethods } from './services/leads/methods.js'
import { StatusesMethods } from './services/statuses/methods.js'
import { UsersMethods } from './services/users/methods.js'
import { Diff } from './utils/Diff.js'
import { ExtractValue } from './utils/ExtractValue.js'

export enum Method {
  BATCH = 'batch',

  CRM_COMPANY_FIELDS = 'crm.company.fields',
  CRM_COMPANY_ADD = 'crm.company.add',
  CRM_COMPANY_UPDATE = 'crm.company.update',
  CRM_COMPANY_GET = 'crm.company.get',
  CRM_COMPANY_LIST = 'crm.company.list',

  CRM_CONTACT_FIELDS = 'crm.contact.fields',
  CRM_CONTACT_ADD = 'crm.contact.add',
  CRM_CONTACT_UPDATE = 'crm.contact.update',
  CRM_CONTACT_GET = 'crm.contact.get',
  CRM_CONTACT_LIST = 'crm.contact.list',

  CRM_DEAL_FIELDS = 'crm.deal.fields',
  CRM_DEAL_ADD = 'crm.deal.add',
  CRM_DEAL_UPDATE = 'crm.deal.update',
  CRM_DEAL_GET = 'crm.deal.get',
  CRM_DEAL_LIST = 'crm.deal.list',

  CRM_LEAD_FIELDS = 'crm.lead.fields',
  CRM_LEAD_ADD = 'crm.lead.add',
  CRM_LEAD_UPDATE = 'crm.lead.update',
  CRM_LEAD_GET = 'crm.lead.get',
  CRM_LEAD_LIST = 'crm.lead.list',

  CRM_STATUS_FIELDS = 'crm.status.fields',
  CRM_STATUS_ADD = 'crm.status.add',
  CRM_STATUS_DELETE = 'crm.status.delete',
  CRM_STATUS_GET = 'crm.status.get',
  CRM_STATUS_LIST = 'crm.status.list',
  CRM_STATUS_UPDATE = 'crm.status.update',

  USER_FIELDS = 'user.fields',
  USER_GET = 'user.get'
}

const LISTABLE_METHODS = [
  Method.CRM_COMPANY_LIST,
  Method.CRM_CONTACT_LIST,
  Method.CRM_DEAL_LIST,
  Method.CRM_LEAD_LIST
] as const

export type ListableMethod = typeof LISTABLE_METHODS[number]
export type GettableMethod = Diff<Method, ListableMethod>

type MethodsMap = {
  readonly [key: string]: {
    readonly type: unknown
    readonly payload: unknown
    readonly params: Record<string, unknown>
  }
}

// @todo Figure out full list of possible values
export type ListParams = {
  readonly start?: number
  readonly order?: { readonly [key: string]: string } // 'ASC' | 'DESC'
  readonly filter?: { readonly [key: string]: string | number }
  readonly select?: readonly string[]
}

/**
 * A mega map of all supported Bitrix methods to their parameters. Used to resolve
 * low-level client method required params and payload types.
 * - `type` — a type that method associated with
 * - `payload` — a payload that method returns
 * - `params` — params that method accepts
 */
export type Methods = MethodsMap & CompaniesMethods & ContactsMethods & DealsMethods & LeadsMethods & StatusesMethods & UsersMethods & {

  readonly [Method.BATCH]: {
    readonly type: unknown
    readonly payload: BatchPayload<unknown>
    readonly params: Commands
  }
}

/**
 * Retrieve specified method data
 */
export type MethodData<M extends Method> = ExtractValue<Methods, M>

/**
 * Retrieve associated with specified method type
 */
export type MethodPayloadType<M extends Method> = MethodData<M>['type']

/**
 * Retrieve specified method payload type
 */
export type MethodPayload<M extends Method> = MethodData<M>['payload']

/**
 * Retrieve specified method params
 */
export type MethodParams<M extends Method> = MethodData<M>['params']
