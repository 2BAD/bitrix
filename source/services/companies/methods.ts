import { ListParams, Method } from '../../methods'
import { GetPayload, ListPayload } from '../../payloads'
import { Fields } from '../common'
import { Company } from './entities'

export interface ContactsMethods {
  readonly [Method.CRM_COMPANY_FIELDS]: {
    readonly type: Company
    readonly payload: GetPayload<Fields>
    readonly params?: {}
  }

  readonly [Method.CRM_COMPANY_ADD]: {
    readonly type: Company
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Company>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.CRM_COMPANY_GET]: {
    readonly type: Company
    readonly payload: GetPayload<Company>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.CRM_COMPANY_LIST]: {
    readonly type: Company
    readonly payload: ListPayload<Company>
    readonly params: ListParams
  }
  readonly [Method.CRM_COMPANY_UPDATE]: {
    readonly type: Company
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, any>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
}
