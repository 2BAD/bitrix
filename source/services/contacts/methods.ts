import { ListParams, Method } from '../../methods'
import { GetPayload, ListPayload } from '../../payloads'
import { Fields } from '../common'
import { Contact } from './entities'

export interface ContactsMethods {
  readonly [Method.CRM_CONTACT_FIELDS]: {
    readonly type: Contact
    readonly payload: GetPayload<Fields>
    readonly params?: Record<string, unknown>
  }

  readonly [Method.CRM_CONTACT_ADD]: {
    readonly type: Contact
    readonly payload: GetPayload<number>
    readonly params: {
      readonly fields: Partial<Contact>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
  readonly [Method.CRM_CONTACT_GET]: {
    readonly type: Contact
    readonly payload: GetPayload<Contact>
    readonly params: {
      readonly id: string
    }
  }
  readonly [Method.CRM_CONTACT_LIST]: {
    readonly type: Contact
    readonly payload: ListPayload<Contact>
    readonly params: ListParams
  }
  readonly [Method.CRM_CONTACT_UPDATE]: {
    readonly type: Contact
    readonly payload: GetPayload<boolean>
    readonly params: {
      readonly id: string
      readonly fields: Record<string, unknown>
      readonly params?: {
        readonly REGISTER_SONET_EVENT: 'Y' | 'N'
      }
    }
  }
}
