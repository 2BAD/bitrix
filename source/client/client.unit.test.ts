import got from 'got'
import nock from 'nock'
import Queue from 'p-queue'
import { Method } from '../methods'
import Client from './index'

const TEST_URI = 'https://test.com/rest'
const TEST_ACCESS_TOKEN = 'test_access_token'
const RESPONSE_200 = 200
const spiedQueueAdd = jest.spyOn(Queue.prototype, 'add')
const client = Client(TEST_URI, TEST_ACCESS_TOKEN)

describe('Client', () => {
  beforeEach(() => {
    spiedQueueAdd.mockClear()
  })

  it('should init', () => {
    expect(() => Client(TEST_URI, TEST_ACCESS_TOKEN)).not.toThrow()
  })

  it('should extend `Got` instance with a specified parameters', () => {
    const spiedGotExtend = jest.spyOn(got, 'extend')

    Client('https://test.com', 'test_token', {https: { rejectUnauthorized: false }})
    expect(spiedGotExtend.mock.calls).toMatchSnapshot()
  })

  it('should expose API methods', () => {
    expect(client).toMatchSnapshot()
  })

  describe('`call`', () => {
    // @todo restore test when issue is resolved (https://github.com/nock/nock/issues/1832)
    it.skip('should add to queue', async () => {
      const payload = {}
      const method = Method.CRM_DEAL_LIST

      nock(TEST_URI)
        .get(`/${method}?access_token=${TEST_ACCESS_TOKEN}`)
        .twice()
        .reply(RESPONSE_200, payload)

      await client.call(method, {})
      await client.call(method, {})

      expect(spiedQueueAdd.mock.calls).toMatchSnapshot()
    })
  })

  describe('`batch`', () => {
    // @todo restore test when issue is resolved (https://github.com/nock/nock/issues/1832)
    it.skip('should add to queue', async () => {
      const payload = {
        result: {
          result: ['done'],
          result_error: []
        }
      }

      const commands = [
        { method: Method.CRM_DEAL_GET },
        { method: Method.CRM_DEAL_LIST }
      ] as const

      nock(TEST_URI)
        // @todo We'd want to use `query` object here as it is much more readable, but nock for some reason
        //       fails to match request when it contains `cmd[someName]`. The issue definitely connected
        //       to the `[]`, since it does not appear when only one bracket present
        .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[0].method}&access_token=${TEST_ACCESS_TOKEN}`)
        .reply(RESPONSE_200, payload)
        .get(`/${Method.BATCH}?cmd%5B0%5D=${commands[1].method}&access_token=${TEST_ACCESS_TOKEN}`)
        .reply(RESPONSE_200, payload)

      await client.batch(commands, 1)

      expect(spiedQueueAdd.mock.calls).toMatchSnapshot()
    })
  })
})
