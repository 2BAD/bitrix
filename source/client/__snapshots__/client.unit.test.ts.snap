// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client \`batch\` should add to queue 1`] = `
Array [
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;

exports[`Client \`call\` should add to queue 1`] = `
Array [
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;

exports[`Client should expose API methods 1`] = `
Object {
  "batch": [Function],
  "call": [Function],
  "list": [Function],
}
`;

exports[`Client should extend \`Got\` instance with a specified parameters 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "user-agent": "@2bad/bitrix",
      },
      "hooks": Object {
        "beforeRequest": Array [
          [Function],
        ],
      },
      "https": Object {
        "rejectUnauthorized": false,
      },
      "prefixUrl": "https://test.com",
      "responseType": "json",
    },
  ],
]
`;
